maas/resource_maas_dns_record_test.go

```diff
@@ -0,0 +1,195 @@
+	return fmt.Sprintf(`
+	resource "maas_dns_record" "test_aaaa_1" {
+	  name   = "%s-1"
+	  type   = "A/AAAA"
+	  data   = %q
+	  domain = "maas"
+	}
+	resource "maas_dns_record" "test_aaaa_2" {
+	  name   = "%s-2"
+	  type   = "A/AAAA"
+	  data   = %q
+	  domain = "maas"
+	}
+	`, recordBaseName, ipAddress, recordBaseName, ipAddress)
+}
+
+func getDNSRecordConfigBasic(name string, recordType string, data string, domain string) string {
+	return fmt.Sprintf(`
+	resource "maas_dns_record" "test" {
+	  name   = "%s"
+	  type   = "%s"
+	  data   = "%s"
+	  domain = "%s"
+	}
+	`, name, recordType, data, domain)
+}
+
+// Check if the DNS record specified exists in MAAS. If not, return an error.
+func testAccMAASDNSRecordCheckExists(rn string, dnsRecord *entity.DNSResource) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		// Check if it exists in state
+		rs, ok := s.RootModule().Resources[rn]
+		if !ok {
+			return fmt.Errorf("resource not found: %s\n %#v", rn, s.RootModule().Resources)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("resource id not set")
+		}
+
+		conn := testutils.TestAccProvider.Meta().(*maas.ClientConfig).Client
+		id, err := strconv.Atoi(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+		gotDNSRecord, err := conn.DNSResource.Get(id)
+		if err != nil {
+			return fmt.Errorf("error getting dns record: %s", err)
+		}
+
+		*dnsRecord = *gotDNSRecord
+
+		return nil
+	}
+}
+
+func testAccMAASDNSRecordCheckDestroy(ipAddress string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testutils.TestAccProvider.Meta().(*maas.ClientConfig).Client
+		for _, rs := range s.RootModule().Resources {
+			// Skip if the resource is not a dns record
+			if rs.Type != "maas_dns_record" {
+				continue
+			}
+			// Convert the resource ID to an integer
+			id, err := strconv.Atoi(rs.Primary.ID)
+			if err != nil {
+				return err
+			}
+
+			// Check if the dns record exists
+			response, err := conn.DNSResource.Get(id)
+			if err == nil {
+				if response != nil && response.ID == id {
+					return fmt.Errorf("dns record still exists: %s", rs.Primary.ID)
+				}
+			}
+
+			// If the error is equivalent to 404 not found, the dns record is destroyed.
+			// Otherwise return the error
+			if !strings.Contains(err.Error(), "404 Not Found") {
+				return err
+			}
+		}
+
+		// Check if the IP address is released
+		ipAllocated, err := isIPAddressAllocated(conn, ipAddress)
+		if err != nil {
+			return fmt.Errorf("error checking if ip address is released: %s", err)
+		}
+		if ipAllocated {
+			return fmt.Errorf("ip address is not released: %s with error: %s", ipAddress, err)
+		}
+		return nil
+	}
+}
+
+// Check if a particular IP address is allocated in MAAS
+func isIPAddressAllocated(conn *client.Client, ipAddress string) (bool, error) {
+	params := &entity.IPAddressesParams{IP: ipAddress}
+	allIPAddresses, err := conn.IPAddresses.Get(params)
```

Comment:
This is not the all IPs since you are using params to filter on the specific IP.

Reply:
If the IP does not exist anymore, you will not receive an error. Instead you will get back an empty slice. This is captured because the for loop is skipped and the function returns false, nil. In case of an error, then you should error out, since it is unexpected. As for the loop, if IP is still allocated, then we should expect that the slice has size 1 and the IP.String() matches the IP in question. In any other case, we also have to error out since this is unexpected.